src/Lib.P2P/Routing/Dht1.cs

```diff
@@ -0,0 +1,380 @@
+                return found;
+
+            // Ask our peers for information on the requested peer.
+            var dquery = new DistributedQuery<Peer>
+            {
+                QueryType = MessageType.FindNode,
+                QueryKey = id,
+                Dht = this,
+                AnswersNeeded = 1
+            };
+            await dquery.RunAsync(cancel).ConfigureAwait(false);
+
+            // If not found, return the closest peer.
+            if (dquery.Answers.Count() == 0) return RoutingTable.NearestPeers(id).FirstOrDefault();
+
+            return dquery.Answers.First();
+        }
+
+        /// <inheritdoc />
+        public Task ProvideAsync(Cid cid, bool advertise = true, CancellationToken cancel = default(CancellationToken))
+        {
+            ContentRouter.Add(cid, Swarm.LocalPeer.Id);
+            if (advertise) Advertise(cid);
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        public async Task<IEnumerable<Peer>> FindProvidersAsync(Cid id,
+            int limit = 20,
+            Action<Peer> action = null,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            var dquery = new DistributedQuery<Peer>
+            {
+                QueryType = MessageType.GetProviders,
+                QueryKey = id.Hash,
+                Dht = this,
+                AnswersNeeded = limit,
+            };
+            if (action != null) dquery.AnswerObtained += (s, e) => action.Invoke(e);
+
+            // Add any providers that we already know about.
+            var providers = ContentRouter
+               .Get(id)
+               .Select(pid =>
+                {
+                    return pid == Swarm.LocalPeer.Id
+                        ? Swarm.LocalPeer
+                        : Swarm.RegisterPeer(new Peer {Id = pid});
+                });
+            foreach (var provider in providers) dquery.AddAnswer(provider);
+
+            // Ask our peers for more providers.
+            if (limit > dquery.Answers.Count()) await dquery.RunAsync(cancel).ConfigureAwait(false);
+
+            return dquery.Answers.Take(limit);
+        }
+
+        /// <summary>
+        ///   Advertise that we can provide the CID to the X closest peers
+        ///   of the CID.
+        /// </summary>
+        /// <param name="cid">
+        ///   The CID to advertise.
+        /// </param>
+        /// <remarks>
+        ///   This starts a background process to send the AddProvider message
+        ///   to the 4 closest peers to the <paramref name="cid"/>.
+        /// </remarks>
+        public void Advertise(Cid cid)
+        {
+            _ = Task.Run(async () =>
+            {
+                var advertsNeeded = 4;
+                var message = new DhtMessage
+                {
+                    Type = MessageType.AddProvider,
+                    Key = cid.Hash.ToArray(),
+                    ProviderPeers = new DhtPeerMessage[]
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)